velox/functions/prestosql/window/LeadLag.cpp

```diff
@@ -0,0 +1,328 @@
+        }
+      }
+    }
+  }
+
+  void setDefaultValue(const VectorPtr& result, int32_t resultOffset) {
+    if (!constantDefaultValue_ && !defaultValueIndex_) {
+      return;
+    }
+
+    // Copy default values into 'result' for rows with invalid offsets or empty
+    // frames.
+
+    if (constantDefaultValue_) {
+      for (auto i = 0; i < rowNumbers_.size(); ++i) {
+        if (rowNumbers_[i] == kNullRow) {
+          result->copy(constantDefaultValue_.get(), resultOffset + i, 0, 1);
+        }
+      }
+    } else {
+      std::vector<vector_size_t> defaultValueRowNumbers;
+      defaultValueRowNumbers.reserve(rowNumbers_.size());
+      for (auto i = 0; i < rowNumbers_.size(); ++i) {
+        if (rowNumbers_[i] == kNullRow) {
+          defaultValueRowNumbers.push_back(partitionOffset_ + i);
+        }
+      }
+
+      if (defaultValueRowNumbers.empty()) {
+        return;
+      }
+
+      partition_->extractColumn(
+          defaultValueIndex_.value(),
+          folly::Range(
+              defaultValueRowNumbers.data(), defaultValueRowNumbers.size()),
+          0,
+          defaultValues_);
+
+      for (auto i = 0; i < defaultValueRowNumbers.size(); ++i) {
+        result->copy(
+            defaultValues_.get(),
+            resultOffset + defaultValueRowNumbers[i] - partitionOffset_,
+            i,
+            1);
+      }
+    }
+  }
+
+  // Index of the 'value' argument.
+  column_index_t valueIndex_;
+
+  // Index of the 'offset' argument if offset is not constant.
+  column_index_t offsetIndex_;
+
+  // Value of the 'offset' if constant.
+  std::optional<int64_t> constantOffset_;
+  bool isConstantOffsetNull_ = false;
+
+  // Index of the 'default_value' argument if default value is specified and not
+  // constant.
+  std::optional<column_index_t> defaultValueIndex_;
+
+  // Constant 'default_value' or null if default value is not constant.
+  VectorPtr constantDefaultValue_;
+
+  const exec::WindowPartition* partition_;
+
+  // Reusable vector of offsets if these are not constant.
+  FlatVectorPtr<int64_t> offsets_;
+
+  // Reusable vector of default values if these are not cosntant.
+  VectorPtr defaultValues_;
+
+  // This offset tracks how far along the partition rows have been output.
+  // This can be used to optimize reading offset column values corresponding
+  // to the present row set in getOutput.
+  vector_size_t partitionOffset_;
+
+  // The Lag function directly writes from the input column to the
+  // resultVector using the extractColumn API specifying the rowNumber mapping
+  // to copy between the 2 vectors. This variable is used for the rowNumber
+  // vector across getOutput calls.
+  std::vector<vector_size_t> rowNumbers_;
+};
+
+template <>
+void LeadLagFunction<true>::setRowNumbersForConstantOffset() {
+  if (isConstantOffsetNull_) {
+    std::fill(rowNumbers_.begin(), rowNumbers_.end(), kNullRow);
+    return;
+  }
+
+  auto constantOffsetValue = constantOffset_.value();
```

Comment:
In general, constantOffsetValue number of rows at the beginning or end are NULL depending on lead or lag. It might be simpler to set consecutive values for rowNumbers for lead and lag both and then just null out the values specifically for each of the functions after that. Wdyt ?

Reply:
@aditi-pandit I'm not sure I understand the suggestion. Any chance you can share the code?

Reply:
This is just a high level pseudocode of what I meant:
void LeadLagFunction<false>::setRowNumbersForConstantOffset() {
  if (isConstantOffsetNull_) {
    std::fill(rowNumbers_.begin(), rowNumbers_.end(), kNullRow);
    return;
  }

  auto constantOffsetValue = constantOffset_.value();
  std::iota(
        rowNumbers_.begin(),
        rowNumbers_.end(),
        partitionOffset_ + constantOffsetValue);
  

  // Figure out how many rows at the end should be NULL.
auto firstNullRowPosition = partition_->numRows() - constantOffsetValue;
auto nullStartPosition = partitionOffset_ + rowNumbers_size() - firstNullRowPostition;
if (nullStartPosition >= 0) {
    std::fill(rowNumbers_.begin() + nullStartPosition, rowNumbers_.end(), kNullRow);
  }
}

Only the last loop changes between lead and lag.

Reply:
@aditi-pandit I see. I tried a bit, but I don't think the result is shorter or easier to read. I'm leaning towards keeping this as is.